Fixed the OOM issue, still have to write tests, release notes, and de…

gmaretic opened the pull request:
…cide how to deal with rows that are too wide

**Goals (and why)**:

Fix the DbKvs sweep OOM issue that arises from the combination of wide rows and large row batch size.

**Implementation Description (bullets)**:

The OOM was happening because the previous implementation created a large multimap before doing the paging. Now we limit the size of the multimap to a predefined maximum (chosen 1M arbitrarily).

**Concerns (what feedback would you like?)**:

Have to decide what to do with rows that are wider than the current GET_RANGE_OF_TS_MAX_BATCH. Current solution is to skip them; alternatively can just process the first GET_RANGE_OF_TS_MAX_BATCH columns. Cannot figure out how to do two pages on the same row as the second set of timestamps seems to just be ignored. 

**Where should we start reviewing?**:

getTimestampsByCell in DbKvs

**Priority (whenever / two weeks / yesterday)**:

Couple of days, as I still have a bunch of things to do before it is ready.

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -576,45 +578,48 @@ private ColumnSelection getColumnSelectionForRange(RangeRequest range) {
         try (ClosableIterator<AgnosticLightResultRow> rangeResults = table.getRange(range, timestamp, maxRows)) {
             while (rows.size() < maxRows && rangeResults.hasNext()) {
-                byte[] rowName = rangeResults.next().getBytes("row_name");
+                byte[] rowName = rangeResults.next().getBlob("row_name");
```
Why was this changed?

jeremyk-91 replied:
`getBytes` was deprecated, replaced with `getBlob` which looks like it returns empty arrays rather than null if the column doesn't exist.

schlosna replied:
see comments below, this may have some perf side effects, would be interested in before/after numbers of this change alone.

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -624,46 +629,57 @@ private ColumnSelection getColumnSelectionForRange(RangeRequest range) {
-            nextRow = RangeRequests.getNextStartRow(range.isReverse(), lastRow);
-            mayHaveMoreResults = rows.size() == maxRows;
+        byte[] nextRow = results.currentRow;
+        boolean mayHaveMoreResults = !results.finishedRange;
```
I would like to see this logic broken out into a place where it's unit testable, as it's quite complex.

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -1313,4 +1329,34 @@ private static Integer getMaxRowsFromBatchHint(@Nullable Integer batchHint) {
+            if (!Arrays.equals(currentRow, cellResult.getBlob("row_name"))) {
+                currentRow = cellResult.getBlob("row_name");
+                storeRow(rowBuffer);
+                rowBuffer.clear();
```
It's very weird to have such a side-effect here. I can see why you'd want it - but maybe we should have this method return a boolean stating whether we stored the row, or alternatively split into methods `rowIsComplete` and `storeRow(cellResult, rowBuffer)`.

gsheasby replied:
Or maybe the abstraction isn't quite right and we need a stateful object to cover the algorithm in `getTimestampsByCell`, to enforce the "store a row and then clear the buffer" requirement.

schlosna commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -624,46 +629,57 @@ private ColumnSelection getColumnSelectionForRange(RangeRequest range) {
+            while (iterator.hasNext() && result.size() + rowBuffer.size() < batchSize) {
+                AgnosticLightResultRow cellResult = iterator.next();
+                result.storePreviousRowIfComplete(cellResult, rowBuffer);
+                Cell cell = Cell.create(cellResult.getBlob("row_name"), cellResult.getBlob("col_name"));
```
why are we using getBlob for columns that are not `BLOB`? We should run some perf tests comparing these for current state on Postgres & Oracle, but from a while back using `cellResult.getBytes("row_name")` was faster & lower overhead than `getBlob`.

gmaretic replied:
That's a good point, I'll look into getting the numbers before this gets a green light. I only changed it in the first place because getBytes specifically says it's deprecated and getBlob should be used instead.

schlosna commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -1313,4 +1329,34 @@ private static Integer getMaxRowsFromBatchHint(@Nullable Integer batchHint) {
+        }
+
+        void storePreviousRowIfComplete(AgnosticLightResultRow cellResult, SetMultimap<Cell, Long> rowBuffer) {
+            if (!Arrays.equals(currentRow, cellResult.getBlob("row_name"))) {
```
see above comment about `getBlob` vs. `getBytes`.

schlosna replied:
Should also pull the row name out to a local to avoid reading twice:

```
byte[] rowName = cellResult.getBytes("row_name");
if (!Arrays.equals(currentRow, rowName)) {
    currentRow = rowName;
```

jeremyk-91 commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -576,45 +578,48 @@ private ColumnSelection getColumnSelectionForRange(RangeRequest range) {
             while (rows.size() < maxRows && rangeResults.hasNext()) {
-                byte[] rowName = rangeResults.next().getBytes("row_name");
+                byte[] rowName = rangeResults.next().getBlob("row_name");
                 if (rowName != null) {
```
Shouldn't need this check (I think) since getBlob returns nonnull

schlosna commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -1313,4 +1329,34 @@ private static Integer getMaxRowsFromBatchHint(@Nullable Integer batchHint) {
     }
+
+    static class ResultWithToken {
+        SetMultimap<Cell, Long> results;
```
nit: `final`

schlosna commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbReadTable.java:
```diff
@@ -62,16 +64,30 @@ public DbReadTable(ConnectionSupplier conns, DbQueryFactory queryFactory) {
+    }
+
+    public ClosableIterator<AgnosticLightResultRow> getAllRows(
+            Iterable<byte[]> rows, ColumnSelection columns, long ts, boolean includeValues, Boolean reverseRows) {
```
taking a `Boolean` arg here feels weird, why not make this a primitive or enum (ASCENDING, DESCENDING, UNDEFINED) for clarity as caller should expressly know desired order?

jeremyk-91 commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbReadTable.java:
```diff
@@ -62,16 +64,30 @@ public DbReadTable(ConnectionSupplier conns, DbQueryFactory queryFactory) {
         }
     }
 
+    private FullQuery addOrdering(Optional<Boolean> reverseRows, FullQuery query) {
```
Could we use Optional above in `getAllRows` or just a Boolean object here? I'd prefer for both to be consistent, if that's feasible.

Perhaps even better: an enum for ordering - undefined, descending, or ascending.

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbReadTable.java:
```diff
@@ -62,16 +64,30 @@ public DbReadTable(ConnectionSupplier conns, DbQueryFactory queryFactory) {
 
     public ClosableIterator<AgnosticLightResultRow> getAllRows(
             Iterable<byte[]> rows, ColumnSelection columns, long ts, boolean includeValues) {
+        return getAllRows(rows, columns, ts, includeValues, null);
```
Passing null here is really ugly - as DavidS said, I'd prefer some kind of enum here.

jeremyk-91 commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -624,46 +629,57 @@ private ColumnSelection getColumnSelectionForRange(RangeRequest range) {
+                .getAllRows(rows, columns, timestamp, false, isReverse)) {
+            ResultWithToken result = new ResultWithToken();
+            SetMultimap<Cell, Long> rowBuffer = HashMultimap.create();
+            while (iterator.hasNext() && result.size() + rowBuffer.size() < batchSize) {
```
Probably worth a note here that `size()` is actually constant time.

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbReadTable.java:
```diff
@@ -62,16 +64,30 @@ public DbReadTable(ConnectionSupplier conns, DbQueryFactory queryFactory) {
 
+    private FullQuery addOrdering(Optional<Boolean> reverseRows, FullQuery query) {
+        String ordering = reverseRows
+                .map(reverse -> " ORDER BY m.row_name" + (reverse ? " DESC" : " ASC") + ", m.col_name, m.ts")
```
nit: the space should probably go in the `" ORDER BY m.row_name"` substring.

gmaretic commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -624,46 +629,57 @@ private ColumnSelection getColumnSelectionForRange(RangeRequest range) {
-                Cell cell = Cell.create(row.getBytes("row_name"), row.getBytes("col_name"));
-                long ts = row.getLong("ts");
-                results.put(cell, ts);
+    private ResultWithToken getTimestampsByCell(
```
That is actually very helpful for the future!

gmaretic commented:
Thanks for the reviews, I have addressed most of it, and added a fix for the wide rows case. There is no hurry to review, as I still have several things to do, so some refactors are expected as well!

gsheasby commented on atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/DbkvsPostgresSweeperTest.java:
```diff
@@ -15,13 +15,113 @@
+        assertThat(results.getNextStartRow().isPresent()).isFalse();
+        assertThat(results.getCellsExamined()).isEqualTo(9);
+        assertThat(results.getCellsDeleted()).isEqualTo(240);
+        resetGetRangeOfTsMaxBatch();
```
`resetGetRangeOfTsMaxBatch` could be done in an `@After` method.

gsheasby commented on atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/DbkvsPostgresSweeperTest.java:
```diff
@@ -15,13 +15,113 @@
+        resetGetRangeOfTsMaxBatch();
+    }
+
+    private void setGetRangeOfTsMaxBatch(long value) {
```
Duplicate code (here until the end of the file). Would be nice to have this code in one place only, e.g. something like `DbKvsTestUtils`.

gmaretic replied:
Makes sense, refactored

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -113,8 +114,12 @@
 
 public class DbKvs extends AbstractKeyValueService {
     private static final Logger log = LoggerFactory.getLogger(DbKvs.class);
+    public static final String ROW = "row_name";
```
Private and public constants should be separate.

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -244,16 +249,16 @@ public void close() {
                 Long overflowId = hasOverflow ? row.getLongObject("overflow") : null;
                 if (overflowId == null) {
-                    Value value = Value.create(row.getBytes("val"), row.getLong("ts"));
+                    Value value = Value.create(row.getBytes("val"), row.getLong(TIMESTAMP));
```
did you consider extracting `"val"`, too?

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -113,8 +114,12 @@
+    public static final long INITIAL_GET_RANGE_OF_TS_BATCH = 1_000_000L;
 
-    private static final long GET_RANGE_OF_TS_MAX_BATCH = 1000000L;
+    private long getRangeOfTsMaxBatch = INITIAL_GET_RANGE_OF_TS_BATCH;
```
perhaps call this `maxRangeOfTimestampsBatchSize`? And "INITIAL" -> "DEFAULT".

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -568,58 +573,91 @@ private ColumnSelection getColumnSelectionForRange(RangeRequest range) {
+     * @return Each RowResult that is returned contains at most getRangeOfTsMaxBatch associated timestamps. It is
+     * guaranteed that all timestamps for a row will be returned in successive RowResults, sorted in ascending order
+     * by column name and timestamp. In addition, the last column and timestamp of each RowResult will be repeated to
+     * ensure sweep does not miss timestamps.
```
This sounds like a hack. Can we do better on the sweep side, e.g. by remembering what the latest timestamp we've seen so far was?

gmaretic replied:
Sure, we can discuss alternative solutions. But in essence they seem equally hacky

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -629,28 +667,20 @@ private ColumnSelection getColumnSelectionForRange(RangeRequest range) {
                 ? ColumnSelection.all() : ColumnSelection.create(range.getColumnNames());
 
-        ResultWithToken results = getTimestampsByCell(table, rows, columns, timestamp, batchSize, range.isReverse());
+        ResultWithToken result = getTimestampsByCell(table, rows, cols, timestamp, batchSize, range.isReverse(), token);
```
This is a lot of parameters. Perhaps it would be good to do a refactor (in a future PR), extracting a method object or parameter object (or both). For now I think we should leave it as-is.

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -629,28 +667,20 @@ private ColumnSelection getColumnSelectionForRange(RangeRequest range) {
         }
-        return SimpleTokenBackedResultsPage.create(nextRow, finalResults, mayHaveMoreResults);
+        List<RowResult<Set<Long>>> finalResults = new ArrayList<>();
+        cellsByRow.entrySet().forEach(entry -> finalResults.add(RowResult.create(entry.getKey(), entry.getValue())));
```
You can do something like:
`List<...> finalResults = cellsByRow.entrySet().forEach(entry -> RowResult.create(...)).collect(Collectors.toList())`.

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -1330,33 +1349,95 @@ private static Integer getMaxRowsFromBatchHint(@Nullable Integer batchHint) {
-        SetMultimap<Cell, Long> results;
-        byte[] currentRow = null;
-        boolean finishedRange = false;
+    static final class ResultWithToken {
```
There's definitely enough logic in here now for this to be a separate class.

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -1330,33 +1349,95 @@ private static Integer getMaxRowsFromBatchHint(@Nullable Integer batchHint) {
+         *  3. In the same column, with higher or equal timestamp (to repeat the last timestamp for sweep)
+         */
+        void moveForward(Token oldToken) {
+            boolean forwarding = oldToken.shouldForward();
```
Perhaps rename `forwarding` to `movingForward` and `shouldForward` to `shouldMoveForward`?

gsheasby replied:
Or skipping / shouldSkip / shouldFinishSkipping.

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:
```diff
@@ -659,25 +689,14 @@ private ResultWithToken getTimestampsByCell(
-            }
+                .getAllRows(rows, columns, timestamp, false, DbReadTable.Order.fromBoolean(isReverse))) {
+            ResultWithToken result = ResultWithToken.create(iterator);
+            result.moveForward(token);
```
It looks like ResultWithToken must be used in this way (calling `create`, `moveForward`, `getBatchOfTimestamps`, and `checkNextEntryAndCreateToken`, in that order), and the object doesn't really make sense until all that has been done.

We could enforce this temporal dependency by moving the other three methods inside `create`, which would then take the iterator, the previous token, and the batch size.

gmaretic replied:
Agree, makes much more sense

gsheasby commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/EquivalenceCountingIterator.java:
```diff
@@ -0,0 +1,85 @@
+    @Override
+    public boolean hasNext() {
+        if (delegate.hasNext()) {
+            if (limit >= totalItems + returnOneIfNextRowIsNew()) {
```
How about an int-returning method `getTotalItems()` that looks like:

```
if (nextRowIsNew()) {
    return totalItems + 1;
} else {
    return totalItems;
}
```

gmaretic replied:
This was so obvious :duck:

gsheasby commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/EquivalenceCountingIterator.java:
```diff
@@ -0,0 +1,85 @@
+
+    @Override
+    public void remove() {
+    // not implemented
```
Throw `NotImplementedException`?

gsheasby commented on docs/source/release_notes/release-notes.rst:
```diff
@@ -40,6 +40,10 @@ develop
          - Change
 
+    *    - |fixed|
+         - Fixed DbKvs sweep OOM issue `#982 <https://github.com/palantir/atlasdb/issues/982>`__ caused by very wide rows. Now uses cell batch size that
```
Incomplete release note!

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/TimestampsByCellResultWithToken.java:
```diff
@@ -0,0 +1,121 @@
+import com.palantir.nexus.db.sql.AgnosticLightResultRow;
+
+final class TimestampsByCellResultWithToken {
+    private static final String ROW = "row_name";
```
These are duplicates of `DbKvs.ROW` etc.

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/TimestampsByCellResultWithToken.java:
```diff
@@ -0,0 +1,121 @@
+     *  2. In a greater column
+     *  3. In the same column, with higher or equal timestamp (to repeat the last timestamp for sweep)
+     */
+    TimestampsByCellResultWithToken moveForward(Token oldToken) {
```
Can these methods (all except `create`) now be private?

gsheasby commented on atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/TimestampsByCellResultWithToken.java:
```diff
@@ -0,0 +1,121 @@
+    private PeekingIterator<AgnosticLightResultRow> iterator;
+
+    final SetMultimap<Cell, Long> entries;
+    boolean mayHaveMoreResults = false;
```
Should probably expose getters for these and make them private, to avoid modification.

gsheasby commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/EquivalenceCountingIterator.java:
```diff
@@ -70,16 +70,16 @@ public T next() {
     @Override
     public void remove() {
-    // not implemented
+        throw new RuntimeException("Method remove() is not implemented.");
```
nit: `UnsupportedOperationException` (java.lang) is more descriptive.

gsheasby commented on atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/DbkvsPostgresTestSuite.java:
```diff
@@ -38,10 +38,10 @@
 @SuiteClasses({
-        DbkvsPostgresKeyValueServiceTest.class,
-        DbkvsPostgresSerializableTransactionTest.class,
+//        DbkvsPostgresKeyValueServiceTest.class,
```
Make sure this is reverted before we merge!

gsheasby commented on atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/sweep/AbstractSweeperTest.java:
```diff
@@ -576,6 +576,27 @@ public void testOom() {
     }
 
+    @Test
+    public void wideRowTest() {
```
How long does this test take to run?

gsheasby commented on atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/AbstractDbKvsKeyValueServiceTest.java:
```diff
@@ -44,27 +44,64 @@ public void resetMaxBatch() {
+                62)) {
+            EquivalenceCountingIterator<RowResult<Set<Long>>> iterator =
+                    new EquivalenceCountingIterator<>(rowResults, 10, SweepTaskRunnerImpl.sameRowEquivalence());
+            assertRowColumnsTimestamps(iterator.next(), 1, ImmutableSet.of(1, 2, 3, 4),
```
I'm concerned that this call is hard to read. Might be good to add a comment that the arguments are: result, expected row key, expected columns, expected timestamps.

gsheasby commented:
Looks like you have checkstyle errors, and need to update the dependency lock.
